overlord/hookstate/hooks.go

```diff
@@ -113,13 +136,36 @@ func (h *gateAutoRefreshHookHandler) Done() error {
 
 	var action snapstate.GateAutoRefreshAction
 	a := ctx.Cached("action")
+
+	// obtain snap lock before manipulating runinhibit lock.
+	var lock *osutil.FileLock
+	if h.refreshAppAwareness {
+		lock, err = snaplock.OpenLock(snapName)
+		if err != nil {
+			return err
+		}
+		if err := lock.Lock(); err != nil {
+			return err
+		}
+		defer lock.Unlock()
+	}
+
+	// default behavior if action is not set
 	if a == nil {
-		// action is unset if the gate-auto-refresh hook exits 0 without
+		// action is not set if the gate-auto-refresh hook exits 0 without
 		// invoking --hold/--proceed; this means proceed (except for respecting
 		// refresh inhibit).
-		if err := snapstate.ProceedWithRefresh(st, snapName); err != nil {
-			return err
+		if h.refreshAppAwareness {
+			// it's inhibited, proceed but preserve inhibit lock.
+			if h.inhibitHint != runinhibit.HintNotInhibited {
```

Comment:
I don't understand this logic.  Two things:

* I would expect that is an invariant that the run inhibit lock is not held outside of an operation on the snap, before it was on snap unlinking, so I'm not sure how the old value could be anything but runinhibit.HintNotInhibited ?
* I would expect that in the default case we return from the hook with the run inhibit lock not held, like it would be right now